Src/Support/Google.Apis.Auth.AspNetCore/GoogleOpenIdConnectExtensions.cs

```diff
@@ -0,0 +1,135 @@
+                options.Scope.Add("profile");
+                // Save the id-token, access-token and refresh-token in the auth properties.
+                options.SaveTokens = true;
+                // Call user configuration.
+                configureOptions(options);
+                // Add event handlers.
+                var userEvents = options.Events;
+                options.Events = new OpenIdConnectEvents
+                {
+                    OnRedirectToIdentityProvider = async ctx =>
+                    {
+                        // Force asking for user consent. This is required to get a refresh-token.
+                        ctx.ProtocolMessage.Prompt = "consent";
+                        // Offline access required to get a refresh-token.
+                        ctx.ProtocolMessage.SetParameter("access_type", "offline");
+                        // Determine if user is already authenticated.
+                        var auth = await ctx.HttpContext.AuthenticateAsync(authenticationScheme);
+                        var authed = auth.Succeeded && !auth.None;
+                        // Handle scopes, with incremental auth if required.
+                        if (ctx.HttpContext.Items.TryGetValue(Consts.HttpContextAdditionalScopeName, out var scope0) && scope0 is string incrementalScope)
+                        {
+                            if (authed)
+                            {
+                                // If user is already authenticated, use incremental auth.
+                                ctx.ProtocolMessage.SetParameter("include_granted_scopes", "true");
+                                ctx.ProtocolMessage.Scope = incrementalScope;
+                            }
+                            else
+                            {
+                                // If user is not authenticated, use standard (non-incremental) auth.
+                                ctx.ProtocolMessage.Scope += $" {incrementalScope}";
+                            }
+                        }
+                        if (authed && auth.Properties.Items.TryGetValue(Consts.ScopeName, out var existingScope))
+                        {
+                            // Pass-through the scopes that are already authorized.
+                            // This is required because all properties are wiped and re-created from this
+                            // auth process. To keep a property requires setting it here; scopes are the only property we need to keep.
+                            ctx.Properties.Items[Consts.ScopeName] = existingScope;
+                        }
+                        // Call user event last so all behaviour can be overridden.
+                        await userEvents.OnRedirectToIdentityProvider(ctx);
+                    },
+                    OnTokenResponseReceived = async ctx =>
+                    {
+                        // TODO: Check this is definitely the correct handler in which to do this.
```

Comment:
Is this a TODO you can do at this point?

Reply:
Thanks, yes, confirmed this is the correct event handler.